tensornetwork/contractors/opt_einsum_paths/path_contractors.py

```diff
@@ -16,15 +16,77 @@
 
 import functools
 import opt_einsum
+import tensornetwork as tn
 from tensornetwork.network import TensorNetwork
-from tensornetwork.network_components import Edge
+from tensornetwork.network_components import Edge, BaseNode
 from tensornetwork.contractors.opt_einsum_paths import utils
-from typing import Any, Optional, Sequence
+from typing import Any, Optional, Sequence, Iterable, Union
 
+#TODO: add return types of functions back once TensorNetwork is gone
 
-def base(net: TensorNetwork,
-         algorithm: utils.Algorithm,
-         output_edge_order: Optional[Sequence[Edge]] = None) -> TensorNetwork:
+
+def _base_nodes(nodes: Iterable[BaseNode],
+                algorithm: utils.Algorithm,
+                output_edge_order: Optional[Sequence[Edge]] = None) -> BaseNode:
+  """Base method for all `opt_einsum` contractors.
+
+  Args:
+    nodes: A collection of connected nodes.
+    algorithm: `opt_einsum` contraction method to use.
+    output_edge_order: An optional list of edges. Edges of the
+      final node in `nodes_set`
+      are reordered into `output_edge_order`;
+      if final node has more than one edge,
+      `output_edge_order` must be pronvided.
+
+  Returns:
+    Final node after full contraction.
+  """
+  nodes_set = set(nodes)
+  tn.check_connected(nodes_set)
+  edges = tn.get_all_nondangling(nodes_set)
+  #output edge order has to be determinded before any contraction
+  #(edges are refreshed after contractions)
+  if output_edge_order is None:
+    output_edge_order = list(
+        (tn.get_all_edges(nodes) - tn.get_all_nondangling(nodes)))
+
+  if set(output_edge_order) != (
+      tn.get_all_edges(nodes) - tn.get_all_nondangling(nodes)):
+    raise ValueError("output edges are not all dangling.")
+
+  for edge in edges:
+    if not edge.is_disabled:  #if its disabled we already contracted it
+      if edge.is_trace():
+        nodes_set.remove(edge.node1)
+        nodes_set.add(tn.contract_parallel(edge))
+
+  if not tn.get_all_nondangling(nodes_set):
+    # There's nothing to contract.
```

Comment:
We should ensure that there is only 1 node in this set too.

Reply:
that's checked above by `tn.check_connected(nodes_set)`. If `nodes_set` had more than one node in it and `get_all_nondangling(nodes_set)` returned an empty `Iterable` then `tn.check_connected(nodes_set)`  above
will have already raised an exception.